### Initialize the Simulation Environment

In [1]:
from sim_class import Simulation

# Initialize the simulation with a specified number of agents
sim = Simulation(num_agents=1)  # For one robot

### Resetting the Simulation

In [2]:
sim.reset(num_agents=1)  # Reset the simulation with the specified number of agents

{'robotId_2': {'joint_states': {'joint_0': {'position': 0.0,
    'velocity': 0.0,
    'reaction_forces': (0.0, 0.0, 0.0, 0.0, 0.0, 0.0),
    'motor_torque': 0.0},
   'joint_1': {'position': 0.0,
    'velocity': 0.0,
    'reaction_forces': (0.0, 0.0, 0.0, 0.0, 0.0, 0.0),
    'motor_torque': 0.0},
   'joint_2': {'position': 0.0,
    'velocity': 0.0,
    'reaction_forces': (0.0, 0.0, 0.0, 0.0, 0.0, 0.0),
    'motor_torque': 0.0}},
  'robot_position': [0.0, 0.0, 0.03],
  'pipette_position': [0.073, 0.0895, 0.1195]}}

### Sending Commands to the Robot



In [3]:
# Check top-right-front corner
sim.reset(num_agents=1)  # Reset the simulation with the specified number of agents
velocity_x = 1
velocity_y = 1
velocity_z = 1
drop_command = 0
actions = [[velocity_x, velocity_y, velocity_z, drop_command]]
sim.run(actions, num_steps=400)

{'robotId_1': {'joint_states': {'joint_0': {'position': -0.18018212830042812,
    'velocity': 0.011489890408608781,
    'reaction_forces': (0.0, 0.0, 0.0, 0.0, 0.0, 0.0),
    'motor_torque': -500.00000000000006},
   'joint_1': {'position': -0.13002333914262054,
    'velocity': 0.001436179847506725,
    'reaction_forces': (0.0, 0.0, 0.0, 0.0, 0.0, 0.0),
    'motor_torque': -500.00000000000006},
   'joint_2': {'position': 0.1700071162742972,
    'velocity': -1.7215376114421329e-06,
    'reaction_forces': (0.0, 0.0, 0.0, 0.0, 0.0, 0.0),
    'motor_torque': 800.0}},
  'robot_position': [0.18018597967177755,
   0.130023017683606,
   0.20000244400881986],
  'pipette_position': [0.2532, 0.2195, 0.2895]}}

In [4]:
# Check top-right-back corner
sim.reset(num_agents=1)  # Reset the simulation with the specified number of agents
velocity_x = -1
velocity_y = 1
velocity_z = 1
drop_command = 0
actions = [[velocity_x, velocity_y, velocity_z, drop_command]]
sim.run(actions, num_steps=400)

{'robotId_2': {'joint_states': {'joint_0': {'position': 0.26001144790330527,
    'velocity': -0.0012399493020369859,
    'reaction_forces': (0.0, 0.0, 0.0, 0.0, 0.0, 0.0),
    'motor_torque': 500.00000000000006},
   'joint_1': {'position': -0.13000790464525305,
    'velocity': 0.0005354119413663154,
    'reaction_forces': (0.0, 0.0, 0.0, 0.0, 0.0, 0.0),
    'motor_torque': -500.00000000000006},
   'joint_2': {'position': 0.17000677860630303,
    'velocity': 1.230356047768519e-05,
    'reaction_forces': (0.0, 0.0, 0.0, 0.0, 0.0, 0.0),
    'motor_torque': 800.0}},
  'robot_position': [-0.26001621250424733,
   0.1300081264677764,
   0.20000197612124446],
  'pipette_position': [-0.187, 0.2195, 0.2895]}}

In [5]:
# Check top-left-front corner
sim.reset(num_agents=1)  # Reset the simulation with the specified number of agents
velocity_x = 1
velocity_y = -1
velocity_z = 1
drop_command = 0
actions = [[velocity_x, velocity_y, velocity_z, drop_command]]
sim.run(actions, num_steps=400)

{'robotId_1': {'joint_states': {'joint_0': {'position': -0.1799991040699111,
    'velocity': 0.00023488698454120476,
    'reaction_forces': (0.0, 0.0, 0.0, 0.0, 0.0, 0.0),
    'motor_torque': -500.00000000000006},
   'joint_1': {'position': 0.2600014396673887,
    'velocity': -7.372167688287386e-05,
    'reaction_forces': (0.0, 0.0, 0.0, 0.0, 0.0, 0.0),
    'motor_torque': 500.00000000000006},
   'joint_2': {'position': 0.17000416332812196,
    'velocity': -7.079825339338202e-06,
    'reaction_forces': (0.0, 0.0, 0.0, 0.0, 0.0, 0.0),
    'motor_torque': 800.0}},
  'robot_position': [0.18000240372159576,
   -0.26000170650838084,
   0.1999992279758396],
  'pipette_position': [0.253, -0.1705, 0.2895]}}

In [6]:
# Check top-left-back corner
sim.reset(num_agents=1)  # Reset the simulation with the specified number of agents
velocity_x = -1
velocity_y = -1
velocity_z = 1
drop_command = 0
actions = [[velocity_x, velocity_y, velocity_z, drop_command]]
sim.run(actions, num_steps=400)

{'robotId_2': {'joint_states': {'joint_0': {'position': 0.26045760697168907,
    'velocity': 0.10999765962966042,
    'reaction_forces': (0.0, 0.0, 0.0, 0.0, 0.0, 0.0),
    'motor_torque': 500.00000000000006},
   'joint_1': {'position': 0.2600020323210818,
    'velocity': -0.0002319970265953906,
    'reaction_forces': (0.0, 0.0, 0.0, 0.0, 0.0, 0.0),
    'motor_torque': 500.00000000000006},
   'joint_2': {'position': 0.17000291645049495,
    'velocity': -0.0002072440726557251,
    'reaction_forces': (0.0, 0.0, 0.0, 0.0, 0.0, 0.0),
    'motor_torque': 800.0}},
  'robot_position': [-0.26045986501428914,
   -0.2600016740798034,
   0.19999894055041878],
  'pipette_position': [-0.1875, -0.1705, 0.2895]}}

In [7]:
# Check bottom-right-front corner
sim.reset(num_agents=1)  # Reset the simulation with the specified number of agents
velocity_x = 1
velocity_y = 1
velocity_z = -1
drop_command = 0
actions = [[velocity_x, velocity_y, velocity_z, drop_command]]
sim.run(actions, num_steps=400)

{'robotId_1': {'joint_states': {'joint_0': {'position': -0.1800056070672756,
    'velocity': -0.0001817948440898788,
    'reaction_forces': (0.0, 0.0, 0.0, 0.0, 0.0, 0.0),
    'motor_torque': -500.00000000000006},
   'joint_1': {'position': -0.13034939925730632,
    'velocity': 0.02212223187040875,
    'reaction_forces': (0.0, 0.0, 0.0, 0.0, 0.0, 0.0),
    'motor_torque': -500.00000000000006},
   'joint_2': {'position': 0.0005004054486476412,
    'velocity': 0.00032048299571364204,
    'reaction_forces': (0.0, 0.0, 0.0, 0.0, 0.0, 0.0),
    'motor_torque': -800.0}},
  'robot_position': [0.17999979589093493,
   0.1303588466474751,
   0.030495361117933363],
  'pipette_position': [0.253, 0.2199, 0.12]}}

In [8]:
# Check bottom-right-back corner
sim.reset(num_agents=1)  # Reset the simulation with the specified number of agents
velocity_x = -1
velocity_y = 1
velocity_z = -1
drop_command = 0
actions = [[velocity_x, velocity_y, velocity_z, drop_command]]
sim.run(actions, num_steps=400)

{'robotId_2': {'joint_states': {'joint_0': {'position': 0.26000567982624795,
    'velocity': 0.00018326250812023383,
    'reaction_forces': (0.0, 0.0, 0.0, 0.0, 0.0, 0.0),
    'motor_torque': 500.00000000000006},
   'joint_1': {'position': -0.130017696380771,
    'velocity': 0.0019882564442310147,
    'reaction_forces': (0.0, 0.0, 0.0, 0.0, 0.0, 0.0),
    'motor_torque': -500.00000000000006},
   'joint_2': {'position': 0.0004838962488522743,
    'velocity': 0.00031152810201852225,
    'reaction_forces': (0.0, 0.0, 0.0, 0.0, 0.0, 0.0),
    'motor_torque': -800.0}},
  'robot_position': [-0.2600037567729493,
   0.13002775374271083,
   0.030477293160401417],
  'pipette_position': [-0.187, 0.2195, 0.12]}}

In [9]:
# Check bottom-left-front corner
sim.reset(num_agents=1)  # Reset the simulation with the specified number of agents
velocity_x = 1
velocity_y = -1
velocity_z = -1
drop_command = 0
actions = [[velocity_x, velocity_y, velocity_z, drop_command]]
sim.run(actions, num_steps=400)

{'robotId_1': {'joint_states': {'joint_0': {'position': -0.18045662637325327,
    'velocity': -0.11002748534412768,
    'reaction_forces': (0.0, 0.0, 0.0, 0.0, 0.0, 0.0),
    'motor_torque': -500.00000000000006},
   'joint_1': {'position': 0.2600004999812214,
    'velocity': -0.00045905284600348574,
    'reaction_forces': (0.0, 0.0, 0.0, 0.0, 0.0, 0.0),
    'motor_torque': 500.00000000000006},
   'joint_2': {'position': 0.0005779855868524066,
    'velocity': 0.0003441607262983139,
    'reaction_forces': (0.0, 0.0, 0.0, 0.0, 0.0, 0.0),
    'motor_torque': -800.0}},
  'robot_position': [0.18046318142592807,
   -0.26000089956038247,
   0.030570836465090945],
  'pipette_position': [0.2535, -0.1705, 0.1201]}}

In [10]:
# Check bottom-left-back corner
sim.reset(num_agents=1)  # Reset the simulation with the specified number of agents
velocity_x = -1
velocity_y = -1
velocity_z = -1
drop_command = 0
actions = [[velocity_x, velocity_y, velocity_z, drop_command]]
sim.run(actions, num_steps=400)

{'robotId_2': {'joint_states': {'joint_0': {'position': 0.2600276914493055,
    'velocity': -0.002312552829850153,
    'reaction_forces': (0.0, 0.0, 0.0, 0.0, 0.0, 0.0),
    'motor_torque': 500.00000000000006},
   'joint_1': {'position': 0.26000205237940205,
    'velocity': -0.0002761664231469496,
    'reaction_forces': (0.0, 0.0, 0.0, 0.0, 0.0, 0.0),
    'motor_torque': 500.00000000000006},
   'joint_2': {'position': 0.0005022216545526984,
    'velocity': 0.0003469825873743042,
    'reaction_forces': (0.0, 0.0, 0.0, 0.0, 0.0, 0.0),
    'motor_torque': -800.0}},
  'robot_position': [-0.26003623004344983,
   -0.2600015943989177,
   0.03049464457536285],
  'pipette_position': [-0.187, -0.1705, 0.12]}}

### All corners

| Corner                  | Position (x, y, z)       |
|-------------------------|--------------------------|
| Top-right-front corner  | 0.2532, 0.2195, 0.2895  |
| Top-right-back corner   | -0.187, 0.2195, 0.2895  |
| Top-left-front corner   | 0.253, -0.1705, 0.2895  |
| Top-left-back corner    | -0.1875, -0.1705, 0.2895|
| Bottom-right-front      | 0.253, 0.2199, 0.12     |
| Bottom-right-back corner| -0.187, 0.2195, 0.12    |
| Bottom-left-front corner| 0.2535, -0.1705, 0.1201 |
| Bottom-left-back corner | -0.187, -0.1705, 0.12   |

